geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/function/BinaryOutputEncoder.scala

```diff
@@ -0,0 +1,226 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.locationtech.geomesa.filter.function
+
+import java.io.OutputStream
+import java.util.Date
+
+import com.typesafe.scalalogging.slf4j.Logging
+import com.vividsolutions.jts.geom.{Geometry, LineString}
+import org.geotools.data.simple.SimpleFeatureCollection
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes.ListAttributeSpec
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+import scala.collection.JavaConversions._
+
+
+object BinaryOutputEncoder extends Logging {
+
+  import org.locationtech.geomesa.filter.function.AxisOrder._
+  import org.locationtech.geomesa.utils.geotools.Conversions._
+
+  case class ValuesToEncode(lat: Float, lon: Float, dtg: Long, track: Option[String], label: Option[String])
+
+  implicit val ordering = new Ordering[ValuesToEncode]() {
+    override def compare(x: ValuesToEncode, y: ValuesToEncode) = x.dtg.compareTo(y.dtg)
+  }
+
+  /**
+   * Encodes a feature collection to bin format. Features are written to the output stream.
+   *
+   * @param fc
+   * @param output
+   * @param dtgField
+   * @param trackIdField
+   * @param labelField
+   * @param latLon
+   * @param axisOrder
+   * @param sort
+   */
+  def encodeFeatureCollection(
+      fc: SimpleFeatureCollection,
+      output: OutputStream,
+      dtgField: String,
+      trackIdField: Option[String],
+      labelField: Option[String] = None,
+      latLon: Option[(String, String)] = None,
+      axisOrder: AxisOrder = LatLon,
+      sort: Boolean = false): Unit = {
+
+    val sft = fc.getSchema
+    val isLineString = sft.getGeometryDescriptor.getType.getBinding == classOf[LineString]
+
+    // validate our input data
+    validateDateAttribute(dtgField, sft, isLineString)
+    validateLabels(trackIdField, labelField, sft)
+    validateLatLong(latLon, sft)
+
+    val sysTime = System.currentTimeMillis
+    val dtgIndex = sft.indexOf(dtgField)
+
+    // get date function based
+    val getDtg: (SimpleFeature) => Long = (f) => {
+      val date = f.getAttribute(dtgIndex).asInstanceOf[Date]
+      if (date == null) sysTime else date.getTime
+    }
+    // for line strings, we need an array of dates corresponding to the points in the line
+    val getLineDtg: (SimpleFeature) => Array[Long] = (f) => {
+      val dates = f.getAttribute(dtgIndex).asInstanceOf[java.util.List[Date]]
+      if (dates == null) Array.empty else dates.map(_.getTime).toArray
```

Comment:
This output format is designed to visualized points by time, so putting all points at the same time wouldn't be useful. Having no date(s) and using systime is a failure condition. They will get skipped them down on line 170.
The other function above uses systime just as an easy error handling condition - it's not consistent between the linestrings and points, as you pointed out, but Longs can't be null so we need some kind of default value.